src/org/pentaho/di/core/namedcluster/NamedClusterManager.java

```diff
@@ -182,4 +189,97 @@ public boolean contains( String clusterName, IMetaStore metastore ) throws MetaS
     return false;
   }
   
+  /**
+   * This method generates the URL from the specific NamedCluster using the
+   * specified scheme.
+   *
+   * @param scheme the name of the scheme to use to create the URL 
+   * @param clusterName the name of the NamedCluster to use to create the URL
+   * @return the generated URL from the specific NamedCluster or null if an error occurs 
+   */
+  public String generateURL( String scheme, String clusterName ) {
+    String clusterURL = null;
+    try {
+      if ( !Const.isEmpty( clusterName ) && !Const.isEmpty( clusterName ) ) {
+        NamedCluster namedCluster = read( clusterName, Spoon.getInstance().getMetaStore() );
+        String ncHostname = null, ncPort = null, ncUsername = null, ncPassword = null;
+        if ( namedCluster != null ) {
+
+          if ( scheme.equals( HadoopSpoonPlugin.HDFS_SCHEME ) ) {
+            ncHostname = namedCluster.getHdfsHost() != null ? namedCluster.getHdfsHost() : "";
+            ncPort = namedCluster.getHdfsPort() != null ? namedCluster.getHdfsPort() : "";
+            ncUsername = namedCluster.getHdfsUsername() != null ? namedCluster.getHdfsUsername() : "";
+            ncPassword = namedCluster.getHdfsPassword() != null ? namedCluster.getHdfsPassword() : "";
+          }
+
+          ncHostname = getVariableSpace().environmentSubstitute( ncHostname );
+          ncPort = getVariableSpace().environmentSubstitute( ncPort );
+          ncUsername = getVariableSpace().environmentSubstitute( ncUsername );
+          ncPassword = getVariableSpace().environmentSubstitute( ncPassword );
+
+          StringBuffer urlString = new StringBuffer( scheme );
```

Comment:
rather than use a StringBuffer, and risk possible encoding issues with usernames/password containing certain characters, let's refactor this code so that it uses something like UrlFileName which we can pass all this same info to, but then ask getURI() to return a safer construction of the URL.